### DISCUSSION TITLE
Is Hermes ready for usage In Game Consoles?

### DISCUSSION THREAD

**Discussion by JailbreakPapa** [facebook/hermes#1296] [General]
Hello,

We've been looking for a JS Engine that we could easily port to consoles(PS5, Xbox), and we came across Hermes. we were able get everything building and somewhat running with the modifications needed, and we are planning to test more with the JS Engine as we finsh porting our HTML/CSS/JS Toolset over to console for UI.

TLDR; Would & Is Hermes ready for usage on restricted platforms like PS5?

**Comment by tmikov** [comment]
We don't know the restrictions or requirements for shipping software in consoles. If you are asking whether it is ready for production use, it has been shipping to more than a billion devices and different OS-es for 5+ years.

Hermes is not perfect for all use cases. For example, our Intl support has gaps (but we have a [plan](https://github.com/facebook/hermes/discussions/1211) for that), performance in CPU-intensive tasks is lower than a JIT (a problem which Static Hermes will partially address), etc, but we are working on it.

Do you have specific concerns?

**Reply by JailbreakPapa** [reply]
The restrictions in place for console platforms like ps5 and Xbox is that the general rule is that JIT is a no-go. 

That's mainly the only concern that SIE & Microsoft have with JavaScript engines on console, so we knew that Engines like V8 & JSC were not going to be a viable option, even with somehow getting them to compile on consoles without JIT. Hermes seems to fill-in that gap since it takes a AOT approach.

**Reply by tmikov** [reply]
Yes, Hermes is used on iOS, where JIT is also prohibited.

**Reply by JailbreakPapa** [reply]
> Yes, Hermes is used on iOS, where JIT is also prohibited.


In that case does Hermes compare pretty well with JSC? 

I could see if I could run some tests with PS5 and test performance while I'm at it.

**Reply by tmikov** [reply]
Hermes is about on par with or slightly faster than "jitless" JSC on many, but not all, benchmarks. It is definitely much faster than JSC on startup and has a small memory footprint.

Obviously I can't make an objective recommendation, but I would say that if you choose Hermes, it should not be as a general purpose JS engine, but it should be for a particular strength (like fast startup, small binary size, memory footprint, etc), or a concrete problem it solves for you. Plus a team that is friendly and responsive to user requests for non-web use cases. We have started bundling non-ES APIs (atob/btoa/TextEncoder) to make our users' lives easier and we plan to continue with a more well-rounded SDK.

Static Hermes, which is the next major release of Hermes, will offer tons of additional non-web improvements like optional native compilation, zero-cost FFI, faster interpreter, builtin support for worker threads, etc.

However if you need a general purpose JS engine, both JSC and v8 can be configured in jitless mode, and will likely always be more up to date with the latest ECMAScript version (though we are currently working hard to improve ours). You can't go wrong with either of them frankly.

**Reply by JailbreakPapa** [reply]
To address us not using JSC or V8 in non JIT mode, it's mainly because the build systems are a pain to port over using the console toolsets, not even doing the underlying platform application itself in C++ (especially with V8/Google's GN system...). 

Hermes wasn't a pain to port over to PS5, and seems to meet our use case in-terms of performance & ECMAScript requirements. the UI system or engine we use doesn't really need every general JS feature that modern sites use today, as it's just being used within a game ui, and nowhere else. 

We are still "open" to using JSC if we do feel like we really need some sort of an general JS engine that is pretty much up-to-date with everything, as I am personally still working on porting that over to PS5 & Xbox, so we can swap them with a simple function call.

But having time at the office today to do some testing of my own, it seems that Hermes was faster than our other fallback JS Engine, so it seems like it will do for now.

**Reply by tmikov** [reply]
Yeah, Hermes is much faster than other engines like QuickJS, Duktape, etc. With Static Hermes native AOT we can be at par or faster than JSC/v8 JIT, depending on the benchmark.
